### DISCUSSION TITLE
How to get RPC/JSON output?

### DISCUSSION THREAD

**Discussion by fritzrehde** [transmission/transmission#3818] [Q&A]
Since the output of commands like `transmission-remote --list` is not intended to be used in shell scripts (the seperators are inconsistent and sometimes just a single space), but rather just to look at, many people suggested using  RPC/JSON in the forum. JSON+jq sounds like a cool solution, but I can't find much documentation on what RPC is and how I can get JSON output from commands like `transmission-remote --list`. Does anyone have example shell scripts? Just adding `--debug`, as per documentation, does not output JSON.

**Comment by Petrprogs** [comment]
What PL do you use?

**Comment by ckerr** [comment]
I don't think there is a clean way to do this right now. If someone wanted to make a PR to add something like `--json` to the command-line options of `transmission-remote` s.t. remote.cc's `processResponse()` just dumped the `response` string to stdout, I'd be happy to review it

**Comment by lavafroth** [comment]
In the meantime, one can query the RPC directly. Here's some bash to get the `name` and `uploadRatio` field of all the current torrents:

```bash
endpoint=localhost:9091/transmission/rpc
header=$(curl --silent $endpoint \
        | grep --perl --only-matching 'X-Transmission-Session-Id: \w+')
curl --header "$header" $endpoint \
     --data '{"method":"torrent-get", "arguments":{"fields":["name", "uploadRatio"]}}'
```

Which outputs:

```json
{"arguments":{"torrents":[{"name":"nixos-24.05-plasma6-x86_64-linux.iso" /* ... */ }]}
```

You may change the fields in the last command to something else. See the table in the [RPC spec](https://github.com/transmission/transmission/blob/main/docs/rpc-spec.md#33-torrent-accessor-torrent-get).

**Comment by valcomm** [comment]
It's already done.
```
$ transmission-remote --help
...
 -j   --json                                              Return RPC response
                                                          as a JSON string
...
$ transmission-remote --authenv -j -l
# a lot of json here
$ transmission-remote --version
transmission-remote 4.0.5 (a6fe2a64aa)
```
